File: python/ch13_agent/agent_ch13/core/events.py
"""运行时事件协议。

Java 类比：``RuntimeEvent`` 类似事件接口，``EventInbox`` 类似
``BlockingQueue<RuntimeEvent>``。后台线程完成任务后不能直接修改 Agent 历史，
而是先把一个强类型事件放入队列，主循环在下一次请求模型前统一取出。
"""

from __future__ import annotations

import json
import threading
from collections import deque
from collections.abc import Mapping
from typing import Protocol

from .messages import ChatMessage, user_message


class RuntimeEvent(Protocol):
    """所有可以注入 Agent Loop 的事件都必须实现的接口。"""

    @property
    def event_id(self) -> str: ...

    @property
    def context_identity(self) -> str | None: ...

    @property
    def idempotency_key(self) -> str | None: ...

    def to_payload(self) -> Mapping[str, object]:
        """把事件转换为只包含 JSON 基础类型的字典。"""


def is_runtime_event(value: object) -> bool:
    """严格检查对象是否实现了事件接口，避免把普通字典误塞进队列。"""
    return (
        hasattr(value, "event_id")
        and isinstance(value.event_id, str)
        and bool(value.event_id.strip())
        and callable(getattr(value, "to_payload", None))
    )


class EventInbox:
    """线程安全的 FIFO 事件队列。"""

    def __init__(self) -> None:
        self._events: deque[RuntimeEvent] = deque()
        self._condition = threading.Condition()

    def publish(self, event: RuntimeEvent) -> None:
        """发布一条事件，并唤醒正在等待的 Loop。"""
        if not is_runtime_event(event):
            raise TypeError("EventInbox 只接受 RuntimeEvent 对象")
        with self._condition:
            self._events.append(event)
            self._condition.notify_all()

    def drain(self, limit: int | None = None) -> tuple[RuntimeEvent, ...]:
        """立即取出当前已就绪事件，保持 FIFO 顺序。"""
        if limit is not None and (isinstance(limit, bool) or limit <= 0):
            raise ValueError("limit 必须是正整数")
        with self._condition:
            count = len(self._events) if limit is None else min(limit, len(self._events))
            return tuple(self._events.popleft() for _ in range(count))

    def wait(self, limit: int | None = None, timeout: float | None = None) -> tuple[RuntimeEvent, ...]:
        """阻塞直到至少有一条事件，然后取出一批事件。"""
        with self._condition:
            if not self._events and not self._condition.wait(timeout):
                return ()
            return self.drain(limit)


def runtime_event_message(event: RuntimeEvent, index: int = 0, total: int = 1) -> ChatMessage:
    """把事件包装成普通 ``user`` 消息，绝不伪装成 tool result。"""
    if not is_runtime_event(event):
        raise TypeError("event 必须是 RuntimeEvent")
    if index < 0 or total <= 0 or index >= total:
        raise ValueError("batch 位置必须满足 0 <= index < total")
    payload = {"runtime_event": dict(event.to_payload()), "batch": {"index": index, "total": total}}
    return user_message(json.dumps(payload, ensure_ascii=False, separators=(",", ":")))

File: typescript/chapters/ch13/src/core/events.ts
// 运行时事件协议：外部异步结果必须先实现 RuntimeEvent 才能进入 EventInbox；
// Agent Loop 批量取出已就绪事件，去重后以普通 user 消息注入历史。
import { userMessage } from "./messages.js";
import type { UserMessage } from "./messages.js";

export interface RuntimeEvent {
  // 事件携带稳定 id 与可选幂等键，供 Loop 去重并将外部结果安全注入下一回合。
  readonly eventId: string;
  readonly contextIdentity?: string;
  readonly idempotencyKey?: string;
  // 把事件序列化为模型可见的纯 JSON 数据；实现方负责稳定字段名。
  toPayload(): Readonly<Record<string, unknown>>;
}

export interface RuntimeEventBatchPosition {
  // 批量注入时记录事件在本批消息中的位置，帮助模型识别这是多条结果中的第几条。
  readonly index: number;
  readonly total: number;
}

// 校验外部对象是否满足 RuntimeEvent 契约；不信任运行时的鸭子类型输入。
export function isRuntimeEvent(value: unknown): value is RuntimeEvent {
  if (typeof value !== "object" || value === null) {
    return false;
  }
  const eventId = Reflect.get(value, "eventId");
  const toPayload = Reflect.get(value, "toPayload");
  const contextIdentity = Reflect.get(value, "contextIdentity");
  const idempotencyKey = Reflect.get(value, "idempotencyKey");
  return (
    typeof eventId === "string" &&
    eventId.trim().length > 0 &&
    typeof toPayload === "function" &&
    (contextIdentity === undefined ||
      (typeof contextIdentity === "string" && contextIdentity.trim().length > 0)) &&
    (idempotencyKey === undefined ||
      (typeof idempotencyKey === "string" && idempotencyKey.trim().length > 0))
  );
}

export class EventInbox {
  // Inbox 是运行时到 Agent Loop 的单向队列；drain 保持 FIFO，并允许一次移交整批事件所有权。
  readonly #events: RuntimeEvent[] = [];
  readonly #waiters: Array<() => void> = [];

  // 发布一条事件并唤醒等待者；只接受已通过 RuntimeEvent 校验的对象。
  publish(event: RuntimeEvent): void {
    if (!isRuntimeEvent(event)) {
      throw new TypeError("EventInbox only accepts RuntimeEvent values");
    }
    this.#events.push(event);
    for (const resolve of this.#waiters.splice(0)) {
      resolve();
    }
  }

  // 取走当前已就绪的事件；limit 缺省为完整 FIFO 批次。
  drain(limit?: number): readonly RuntimeEvent[] {
    if (limit !== undefined && (!Number.isInteger(limit) || limit <= 0)) {
      throw new Error("limit must be a positive integer or undefined");
    }
    const count = limit === undefined ? this.#events.length : Math.min(limit, this.#events.length);
    return Object.freeze(this.#events.splice(0, count));
  }

  // 阻塞等待队列非空后取走一批事件；由 Agent Loop 在等待后台结果时调用。
  async wait(limit?: number): Promise<readonly RuntimeEvent[]> {
    while (this.#events.length === 0) {
      await new Promise<void>((resolve) => this.#waiters.push(resolve));
    }
    return this.drain(limit);
  }
}

// 事件以普通 user 消息进入历史，但不伪装成 tool result，也不携带 tool_call_id。
// batch 只在一次 drain 注入多条事件时使用；未传时按单条事件处理。
export function runtimeEventMessage(
  event: RuntimeEvent,
  batch?: RuntimeEventBatchPosition,
): UserMessage {
  if (!isRuntimeEvent(event)) {
    throw new TypeError("event must be a RuntimeEvent");
  }
  const position = batch === undefined ? { index: 0, total: 1 } : batch;
  if (
    !Number.isInteger(position.index) ||
    position.index < 0 ||
    !Number.isInteger(position.total) ||
    position.total <= 0 ||
    position.index >= position.total
  ) {
    throw new TypeError("batch position must satisfy 0 <= index < total");
  }
  return userMessage(
    JSON.stringify(
      { runtime_event: event.toPayload(), batch: position },
      (_key, value: unknown) => value,
    ),
  );
}
